src/core/qgssnappingutils.cpp

```diff
@@ -330,6 +330,57 @@ QgsPointLocator::Match QgsSnappingUtils::snapToMap( const QgsPointXY &pointMap,
 }
 
 
+QgsFeatureIds QgsSnappingUtils::isMatchAVisibleLayer( QgsVectorLayer *layer )
+{
+
+  QgsFeatureIds fids;
+
+    if ( layer )
+    {
+        if (!mEnableSnappingForInvisibleFeature) {
+      bool filter = false;
+      QgsRenderContext context( QgsRenderContext::fromMapSettings( mMapSettings ) );
+      context.expressionContext() << QgsExpressionContextUtils::layerScope( layer );
+      std::unique_ptr< QgsFeatureRenderer > renderer( layer->renderer() ? layer->renderer()->clone() : nullptr );
+      if ( renderer )
+      {
+        // setup scale for scale dependent visibility (rule based)
+        renderer->startRender( context, layer->fields() );
+        filter = renderer->capabilities() & QgsFeatureRenderer::Filter;
+      }
+
+
+      QgsFeatureIterator fi = layer->getFeatures();
+
+      QgsFeature feat;
+      while ( fi.nextFeature( feat ) )
+      {
+        context.expressionContext().setFeature( feat );
+
+        if ( !(filter && !renderer->willRenderFeature( feat, context ) ) )
+            fids.insert( feat.id() );
+      }
+
+      if ( renderer )
+      {
+        renderer->stopRender( context );
+      }
+        }
+        else {
```

Comment:
I don't think this else branch is needed - if the invisible skipping is disabled, you could return an empty list here. That'll be handled correctly by the request in QgsPointLocator::rebuildIndex and avoid the expensive full layer iteration performed here.

Reply:
done